design/824-add-nchains-option-to-mcmc/discussion.qmd

```diff
@@ -0,0 +1,1202 @@
+`rjags` supports multiple chains via the `n.chains` argument to `jags.model` and related functions.
+All that it appears we need to do is to add a similar parameter to `mcmc` and we have what we need.
+
+And indeed, this will/should work. But it will not be effective. The reason is that, to test
+convergence effectively, different chains estimating the same model should start from different
+initial positions. If these positions are extreme, so much the better, as it supports the assumption
+that the chains will converge (to the same stable state) regardless of their starting positions.
+
+This is where our difficulties begin. For each model class, `crmPack` provides *constant* starting
+values for the MCMC chain(s). For example, this is the `init` slot for `LogisticLogNormal`:
+
+```{r}
+#| eval: FALSE
+#| label: example-init-slot
+
+init <- function() {
+  list(theta = c(0, 1))
+}
+```
+
+Clearly, what values can be considered extreme depend on the prior distribution of the model. These
+are provided elsewhere in the model object. (For example in the `params` slot of `ModelLogNormal`.)
+This is where our difficulties deepen:
+
+> There is no direct way for a slot of an S4 object to access the value in another slot *of the same
+> object*.
+
+## Is there a need for an `n.chains` argument to `mcmc`?
+
+An `n.chains` argument to `mcmc` might be useful, but I don't believe it is necessary. My reason for
+saying so is the way in which chains are initialised in JAGS. When using a single chain, users
+simply supply a `list` containing the necessary information to initialise the chain. To use multiple
+chains, the user supplies a `list` of such `list`s and sets `n.chains` to the length of the outer
+`list`. It would be a simple matter for `crmPack` to set `n.chains` for `jags.model` on the basis of
+the length of the list of user supplied `inits`.
+
+Of course, if the user wanted `crmPack` to automate the creation of the model `inits`, then
+`n.chains` would be useful. However, as I shall demonstrate later in this document, I think there
+are better ways to do this.
+
+```{r}
+#| label: list-of-lists-demo
+#| eval: FALSE
+#| echo: FALSE
+model <- rjags::jags.model(
+  file = textConnection(
+    "model {
+      theta ~ dmnorm(mean, prec)
+      alpha0 <- theta[1]
+      alpha1 <- theta[2]
+    }"
+  ),
+  data = list(
+    mean = c(-0.85, 1.00),
+    prec = matrix(c(1.333, 0.667, 0.667, 1.333), nrow = 2)
+  ),
+  inits = list(
+    list(theta = c(1, 0)),
+    list(theta = c(0, 1))
+  ),
+  quiet = FALSE,
+  n.adapt = 0,
+  n.chains = 2
+)
+update(model, n.iter = 500)
+samples <- rjags::jags.samples(
+  model = model,
+  variable.names = c("alpha0", "alpha1"),
+  n.iter = 500,
+  thin = 2
+)
+samples
+```
+
+# Simplistic support for multiple chains
+
+By modifying the
+function`in a`Model`'s`init`slot to take a parameter that defines the number of chains required, multiple`inits\`
+can be easily supported without breaking existing code:
```

Comment:
broken text? or ok?

Reply:
Broken text.  Now fixed.  If you agree that using a separate `initialise` function is the way to go, then this section (and the following one) _could_ be deleted.  On the other hand, keeping them might provide useful context in the future...